src/sdk/pynni/nni/nas/pytorch/search_space_zoo/enas_cell.py

```diff
@@ -0,0 +1,254 @@
+    def __init__(self, node_name, prev_node_names, channels):
+        super().__init__(node_name)
+        self.cell_x = Cell(node_name + "_x", prev_node_names, channels)
+        self.cell_y = Cell(node_name + "_y", prev_node_names, channels)
+
+    def forward(self, prev_layers):
+        out_x, mask_x = self.cell_x(prev_layers)
+        out_y, mask_y = self.cell_y(prev_layers)
+        return out_x + out_y, mask_x | mask_y
+
+
+class Calibration(nn.Module):
+    def __init__(self, in_channels, out_channels):
+        super().__init__()
+        self.process = None
+        if in_channels != out_channels:
+            self.process = StdConv(in_channels, out_channels)
+
+    def forward(self, x):
+        if self.process is None:
+            return x
+        return self.process(x)
+
+
+class ENASReductionLayer(nn.Module):
+    """
+    builtin EnasReductionLayer.
+
+    Parameters
+    ---
+    in_channels_pp: int
+        the number of previous previous layer's output channels
+    in_channels_p: int
+        the number of previous layer's output channels
+    out_channels: int
+        output channels of this layer
+    """
+    def __init__(self, in_channels_pp, in_channels_p, out_channels):
+        super().__init__()
+        self.reduce0 = FactorizedReduce(in_channels_pp, out_channels, affine=False)
+        self.reduce1 = FactorizedReduce(in_channels_p, out_channels, affine=False)
+
+    def forward(self, pprev, prev):
+        return self.reduce0(pprev), self.reduce1(prev)
+
+
+class ENASMicroLayer(nn.Module):
+    """
+    Builtin EnasMicroLayer. Micro search designs only one building block whose architecture is repeated
+    throughout the final architecture. A cell has ``num_nodes`` nodes and searches the topology and
+    operations among them in RL way. The first two nodes in a layer stand for the outputs from previous
+    previous layer and previous layer respectively. For the following nodes, the controller chooses
+    two previous nodes and applies two operations respectively for each node. Nodes that are not served
+    as input for any other node are viewed as the output of the layer. If there are multiple output nodes,
+    the model will calculate the average of these nodes as the layer output. Every node's output has ``out_channels``
+    channels so the result of the layer has the same number of channels as each node.
+
+    Parameters
+    ---
+    cells_list: list
+        the list contains all cells in model
+    num_nodes: int
+        the number of nodes contained in this layer
+    in_channles_pp: int
+        the number of previous previous layer's output channels
+    in_channels_p: int
+        the number of previous layer's output channels
+    out_channels: int
+        output channels of this layer
+    reduction: bool
+        is previous layer a reduction layer
+    """
+    def __init__(self, cells_list, num_nodes, in_channels_pp, in_channels_p, out_channels, reduction):
```

Comment:
I do not think passing in the list of cells is proper here. Maybe we need another way to hide `ENASReductionLayer`.